backend/executor.go

```diff
@@ -202,22 +228,76 @@ func (g *grpcExecutor) GetWorkItems(req *protos.GetWorkItemsRequest, stream prot
 	callback := g.onWorkItemConnection
 	if callback != nil {
 		if err := callback(stream.Context()); err != nil {
-			message := fmt.Sprint("unable to establish work item stream at this time: ", err)
+			message := "unable to establish work item stream at this time: " + err.Error()
 			g.logger.Warn(message)
 			return status.Errorf(codes.Unavailable, message)
 		}
 	}
 
+	// Collect all pending activities on this stream
+	// Note: we don't need sync.Map's here because access is only on this thread
+	pendingActivities := make(map[string]struct{})
+	pendingActivityCh := make(chan string, 1)
+	pendingOrchestrators := make(map[string]struct{})
+	pendingOrchestratorCh := make(chan string, 1)
+	defer func() {
+		// If there's any pending activity left, remove them
+		for key := range pendingActivities {
+			g.logger.Debug("cleaning up pending activity", key)
+			p, ok := g.pendingActivities.LoadAndDelete(key)
+			if ok {
+				pending := p.(*activityExecutionResult)
+				close(pending.complete)
+			}
+		}
+		for key := range pendingOrchestrators {
+			g.logger.Debug("cleaning up pending orchestrator", key)
+			p, ok := g.pendingOrchestrators.LoadAndDelete(api.InstanceID(key))
+			if ok {
+				pending := p.(*ExecutionResults)
+				close(pending.complete)
+			}
+		}
+	}()
+
 	// The worker client invokes this method, which streams back work-items as they arrive.
 	for {
 		select {
 		case <-stream.Context().Done():
-			g.logger.Infof("work item stream closed")
+			g.logger.Info("work item stream closed")
 			return nil
-		case wi := <-g.workItemQueue:
+		case wi, ok := <-g.workItemQueue:
+			if !ok {
+				continue
+			}
+			switch x := wi.Request.(type) {
+			case *protos.WorkItem_OrchestratorRequest:
+				key := x.OrchestratorRequest.GetInstanceId()
+				pendingOrchestrators[key] = struct{}{}
+				p, ok := g.pendingOrchestrators.Load(api.InstanceID(key))
+				if ok {
+					p.(*ExecutionResults).pending = pendingOrchestratorCh
+				}
+				g.logger.Debugf("pending orchestrators after add %s: %#v\n", key, pendingOrchestrators)
```

Comment:
I think these debug logs could be removed @kaibocai 
I had left the logs in the code I shared with you just to show the flow of data! These logs may be even too verbose for a debug-level logging

Reply:
I removed logs here and just left a few logs on tracking the newly added logic.